crates/burn-core/src/lr_scheduler/step.rs

```diff
@@ -0,0 +1,178 @@
+use burn_tensor::backend::Backend;
+
+use crate as burn;
+
+use super::LrScheduler;
+use crate::{config::Config, LearningRate};
+
+/// The configuration for create a [step learning rate scheduler](StepLrScheduler).
+///
+/// This scheduler returns the learning rate `initial_lr` from the start, and keeps doing so until
+/// the same value has been given for `step_size` times. Then it multiplies the learning rate by
+/// `gamma` before repeating the process.
+///
+/// ## Notes
+///
+/// The [step](StepLrScheduler::step) method of the scheduler panics if it is called more than
+/// `i32::MAX + 1` times.
+#[derive(Config)]
+pub struct StepLrSchedulerConfig {
+    // The learning rate at the initial step.
+    initial_lr: LearningRate,
+    // The number of iterations over which the learning rate remains unchanged before the next
+    // update.
+    step_size: usize,
+    /// The factor by which the learning rate is multiplied with each update. Default: 0.1.
+    #[config(default = 0.1)]
+    gamma: f64,
+}
+
+impl StepLrSchedulerConfig {
+    /// Initializes a [step learning rate scheduler](StepLrScheduler).
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if any field is out of the acceptable range.
+    pub fn init(&self) -> Result<StepLrScheduler, String> {
+        // `initial_lr` and `gamma` are not checked because atypical values such as zero and
+        // negative values might be useful in some cases like debugging (e.g.,
+        // https://datascience.stackexchange.com/q/89518).
+        if self.step_size == 0 {
+            return Err("Step size must be greater than 0".into());
+        }
+
+        Ok(StepLrScheduler {
+            init_lr: self.initial_lr,
+            step_size: self.step_size,
+            gamma: self.gamma,
+            iter_idx: -1,
+        })
+    }
+}
```

Comment:
Is there a particular reason why the initialization uses `Result` and returns and `Err` instead of assertions/panics for the value checks?

Just asking because the pattern is different from the other schedulers.

Also, regarding `gamma` and `init_lr`, I think a warnings would be a good idea as you suggested.

Reply:
Thank you for your feedback.

I do have 2 reasons for using `Result`. One is for leaving callers the choice of how to handle the error. The other one is that returning `Result` seems acceptable since I find there are some other structs (not LR schedulers) doing so too. But neither of the reasons is a strong one and it does break the consistency, so if it's better to stick with the assertion way, I will fix that part.

As for the warning part, it seems that warnings should be printed by `log::warn!`, but I find a little issue with it. This method is a noop unless users explicitly create an application logger and call `.with_application_logger(None)` before building the learner. This is because the built-in logger is created in `LearnerBuilder::build`, which is unavailable when the LR scheduler is initialized.

I tried the MNIST example, however, and found that actually there were some log (`AdapterInfo` from cubecl_wgpu and `New execution plan 0` from burn_fusion) being discarded due to being logged at a stage too early, so letting users create the logger may not be that bad.

Reply:
I actually agree for `Result`, just that we would have to change the behavior of existing schedulers for consistency 😅 so for this PR addition I would keep it consistent with others, and possibly make the change to `Result` for all schedulers in a follow-up PR.

Regarding the warning, good catch! I'll think about what we should do, but the warning at the optimizer init level should be valid. If we're not logging because it hasn't been initialized, that's another issue to fix somewhere else I believe.

Reply:
Yeah, one thing at a time is better. I will replace `Result` with an assertion and add warnings by `log::warn!` for initial LR and gamma, and leave the issues with returning value of `.init()` and logging to future PRs.

Reply:
Perfect. We should open issues for the follow-up works 👍